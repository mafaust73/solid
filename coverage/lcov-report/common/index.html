<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for common</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> common
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">95.05% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>192/202</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">80.26% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>61/76</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">94.34% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>50/53</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">95.36% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>185/194</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="bspNode.js"><a href="bspNode.js.html">bspNode.js</a></td>
	<td data-value="86.49" class="pic high"><div class="chart"><div class="cover-fill" style="width: 86%;"></div><div class="cover-empty" style="width:14%;"></div></div></td>
	<td data-value="86.49" class="pct high">86.49%</td>
	<td data-value="74" class="abs high">64/74</td>
	<td data-value="73.91" class="pct medium">73.91%</td>
	<td data-value="46" class="abs medium">34/46</td>
	<td data-value="78.57" class="pct medium">78.57%</td>
	<td data-value="14" class="abs medium">11/14</td>
	<td data-value="87.14" class="pct high">87.14%</td>
	<td data-value="70" class="abs high">61/70</td>
	</tr>

<tr>
	<td class="file empty" data-value="common.js"><a href="common.js.html">common.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="csg.js"><a href="csg.js.html">csg.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="37" class="abs high">37/37</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="35" class="abs high">35/35</td>
	</tr>

<tr>
	<td class="file high" data-value="plane.js"><a href="plane.js.html">plane.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="48" class="abs high">48/48</td>
	<td data-value="86.36" class="pct high">86.36%</td>
	<td data-value="22" class="abs high">19/22</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="48" class="abs high">48/48</td>
	</tr>

<tr>
	<td class="file high" data-value="polygon.js"><a href="polygon.js.html">polygon.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file high" data-value="vector.js"><a href="vector.js.html">vector.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="29" class="abs high">29/29</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="15" class="abs high">15/15</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="29" class="abs high">29/29</td>
	</tr>

<tr>
	<td class="file high" data-value="vertex.js"><a href="vertex.js.html">vertex.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sun Jul 01 2018 12:42:36 GMT+0200 (CEST)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
<script src="../block-navigation.js"></script>
</body>
</html>
