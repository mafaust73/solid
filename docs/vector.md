# Vector

Create a vector using `new Vector(x, y, z)`, e.g. `new Vector(1, 0, 0)`;

## Operations

All operations are immutable, i.e. returning always a new vector.

| Operation                        | Method                               |
| -------------------------------- | ------------------------------------ |
| a + b                            | a.plus(b)                            |
| a - b                            | a.minus(b)                           |
| a \* 7                           | a.times(7)                           |
| a / 7                            | a.dividedBy(7)                       |
| a ⋅ b                            | a.dot(b)                             |
| a ⨯ b                            | a.cross(b)                           |
| -a                               | a.negated()                          |
| Interpolate between vectors      | a.lerp(b, t) with 0 ≤ t ≤ 1          |
| Length of a vector               | a.length()                           |
| Normalize a vector to length = 1 | a.unit()                             |
| Rotate around y-axis             | a.rotateY(alpha) with 0 ≤ alpha ≤ 2π |
| Return a digital copy            | a.clone()                            |

## Constants

The library provides constants for

| Constant          | Value                                    |
| ----------------- | ---------------------------------------- |
| ZERO              | new Vector(0, 0, 0)                      |
| ONE               | new Vector(1, 1, 1)                      |
| UNIT_X            | new Vector(1, 0, 0)                      |
| UNIT_Y            | new Vector(0, 1, 0)                      |
| UNIT_Z            | new Vector(0, 0, 1)                      |
| POSITIVE_INFINITY | new Vector(Infinity, Infinity, Infinity) |
| NEGATIVE_INFINITY | new Vector(Infinity, Infinity, Infinity) |
