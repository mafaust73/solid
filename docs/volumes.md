# Volumes

## Simple

### Cube

A cube is described by:

- center : Vector
- size : Vector.

```javascript
import { Cube, ZERO, ONE } from "solid";

let center = ZERO;
let size = ONE;

let cube = new Cube(center, size);
```

![Cube](cube.png)

### Sphere

A sphere is described by:

- center : Vector
- radius : Number
- resolution : Integer  
  With `resolution` you can control the granularity of the approximation.

```javascript
import { Sphere, ZERO, ONE } from "solid";

let center = ZERO;
let radius = 1;
let resolution = 10;

let sphere = new Sphere(center, radius, resolution);
```

![Sphere](sphere.png)

### Cylinder

A cylinder is described by:

- start : Vector
- end : Vector
- radius : Number
- resolution : Integer  
  With `resolution` you can control the granularity of the approximation.

```javascript
import { Cylinder, ZERO, ONE, UNIT_Y } from "solid";

let start = ZERO;
let end = UNIT_Y;
let radius = 1;
let resolution = 10;

let cylinder = new Cylinder(start, end, radius, resolution);
```

![Cylinder](cylinder.png)

### Cone

A cone is described by:

- start : Vector
- end : Vector
- radiusStart : Number
- radiusEnd : Number
- closed : Boolean
- resolution : Integer  
  With `resolution` you can control the granularity of the approximation.

```javascript
import { Cone, ZERO, ONE, UNIT_Y } from "solid";

let start = ZERO;
let end = UNIT_Y;
let radiusStart = 1;
let radiusEnd = 0.2;
let resolution = 10;

let cone = new Cone(start, end, radiusStart, radiusEnd, resolution);
```

![Cone](cone.png)

## Complex

### Lathe

Creates a rotational object:

- path : Path the path in the xy plane
- resolution : Integer the path resolution
- slices Integer the number of slices around the rotational axis

```javascript
import { Lathe, Path } from "solid";

let path = new Path()...;
let slices = 32;

let lathe = new Lathe(path, resolution, slices);
```

![Lathe](lathe.png)

### Twist

Creates a rotational object:

- path : Path the path in the xz plane
- aspect : Number the amount of twist per height unit
- resolution : Integer the path resolution
- stacks Integer the number of stacks along the y-axis

![Twist](twist.png)

### Path

Paths are building the base for many volumes. Paths are build in an abstract U-V-Space and translated according to the given axis.

The following image shows a simple path for creating a glas object.
![path](path.svg)

A path is created using a flow builder pattern:

```javascript
let path = new Path()
  .moveTo(0, 0)
  .lineTo(10, 0)
  .lineTo(10, 2);
```

| Operation    | Description                       |
| ------------ | --------------------------------- |
| moveTo(u, v) | Must be the first item of a path. |
| lineTo(u, v) | Create a linear segment           |

TODO: arc and bezier paths

## Combine geometries

Combine two geometries by union, subtract and intersect.

```javascript
import { Cube, Sphere, ZERO, ONE } from "solid";
let cube = new Cube(ZERO, ONE);
let sphere = new Sphere(ZERO, 1.2, 32);

let union = cube.union(sphere);
let subtract = cube.subtract(sphere);
let intersect = cube.intersect(sphere);
```

| union               | subtract                  | intersect                   |
| ------------------- | ------------------------- | --------------------------- |
| ![union](union.png) | ![subtract](subtract.png) | ![intersect](intersect.png) |
