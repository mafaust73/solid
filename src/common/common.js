export * from "./csg";
export * from "./plane";
export * from "./polygon";
export * from "./vector";
export * from "./vertex";
