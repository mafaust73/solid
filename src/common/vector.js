export class Vector {
  constructor(x, y, z) {
    this.x = x || 0;
    this.y = y || 0;
    this.z = z || 0;
  }

  min(v) {
    return new Vector(Math.min(this.x, v.x), Math.min(this.y, v.y), Math.min(this.z, v.z));
  }
  max(v) {
    return new Vector(Math.max(this.x, v.x), Math.max(this.y, v.y), Math.max(this.z, v.z));
  }

  clone() {
    return new Vector(this.x, this.y, this.z);
  }

  negated() {
    return new Vector(-this.x, -this.y, -this.z);
  }

  plus(a) {
    return new Vector(this.x + a.x, this.y + a.y, this.z + a.z);
  }

  minus(a) {
    return new Vector(this.x - a.x, this.y - a.y, this.z - a.z);
  }

  times(a) {
    return new Vector(this.x * a, this.y * a, this.z * a);
  }

  dividedBy(a) {
    return new Vector(this.x / a, this.y / a, this.z / a);
  }

  dot(a) {
    return this.x * a.x + this.y * a.y + this.z * a.z;
  }

  lerp(a, t) {
    let dx = a.x - this.x;
    let dy = a.y - this.y;
    let dz = a.z - this.z;
    return new Vector(this.x + t * dx, this.y + t * dy, this.z + t * dz);
  }

  length() {
    return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
  }

  unit() {
    return this.dividedBy(this.length());
  }

  cross(a) {
    return new Vector(this.y * a.z - this.z * a.y, this.z * a.x - this.x * a.z, this.x * a.y - this.y * a.x);
  }

  rotateY(alpha) {
    let c = Math.cos(alpha);
    let s = Math.sin(alpha);

    return new Vector(this.x * c + this.z * s, this.y, -this.x * s + this.z * c);
  }
}

export const NEGATIVE_INFINITY = new Vector(-Infinity, -Infinity, -Infinity);
export const POSITIVE_INFINITY = new Vector(Infinity, Infinity, Infinity);

export const ZERO = new Vector(0, 0, 0);
export const ONE = new Vector(1, 1, 1);

export const UNIT_X = new Vector(1, 0, 0);
export const UNIT_Y = new Vector(0, 1, 0);
export const UNIT_Z = new Vector(0, 0, 1);
