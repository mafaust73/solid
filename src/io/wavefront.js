import fs from "fs";

/**
 * Export as Alias Wavefront obj file.
 */
export class Wavefront {
  /**
   * Encode csg node to obj.
   * @param {CSG} csg the node
   * @returns {String} the obj file
   */
  write(csg) {
    let vertices = "";
    let normals = "";
    let faces = "";

    let offset = 0;
    csg.polygons.forEach(p => {
      let a = p.vertices[0].position;
      let b = p.vertices[1].position;

      vertices += `v ${a.x} ${a.y} ${a.z}\n`;
      vertices += `v ${b.x} ${b.y} ${b.z}\n`;

      a = p.vertices[0].normal;
      b = p.vertices[1].normal;
      normals += `vn ${a.x} ${a.y} ${a.z}\n`;
      normals += `vn ${b.x} ${b.y} ${b.z}\n`;

      for (let i = 2; i < p.vertices.length; i++) {
        let c = p.vertices[i].position;
        vertices += `v ${c.x} ${c.y} ${c.z}\n`;
        c = p.vertices[i].normal;
        normals += `vn ${c.x} ${c.y} ${c.z}\n`;

        let ax = 1 + offset;
        let bx = 1 + offset + i - 1;
        let cx = 1 + offset + i;
        faces += `f ${ax}//${ax} ${bx}//${bx} ${cx}//${cx}\n`;
      }

      offset += p.vertices.length;
    });

    return vertices + normals + faces;
  }
}
