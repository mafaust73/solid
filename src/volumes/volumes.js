export { Cube } from "./cube";
export { Cylinder } from "./cylinder";
export { Sphere } from "./sphere";
export { Cone } from "./cone";
export { Lathe } from "./lathe";
export { Twist } from "./twist";
