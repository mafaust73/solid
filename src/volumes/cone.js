import { CSG, Vector, Vertex, Polygon } from "../common/common";

export class Cone extends CSG {
  constructor(start, end, radiusStart, radiusEnd, closed, resolution) {
    super();

    let ray = end.minus(start);
    let axisZ = ray.unit();
    let isY = Math.abs(axisZ.y) > 0.5;
    let axisX = new Vector(isY ? 1 : 0, !isY ? 1 : 0, 0).cross(axisZ).unit();
    let axisY = axisX.cross(axisZ).unit();

    let vxStart = new Vertex(start, axisZ.negated());
    let vxEnd = new Vertex(end, axisZ.unit());

    if (radiusStart === 0) {
      for (let i = 0; i < resolution; i++) {
        let t0 = (i / resolution) * 2 * Math.PI;
        let t1 = (((i + 1) % resolution) / resolution) * 2 * Math.PI;

        let d0 = axisX.times(Math.cos(t0)).plus(axisY.times(Math.sin(t0)));
        let p0 = end.plus(d0.times(radiusEnd));
        let d1 = axisX.times(Math.cos(t1)).plus(axisY.times(Math.sin(t1)));
        let p1 = end.plus(d1.times(radiusEnd));
        this.polygons.push(new Polygon([vxStart, new Vertex(p0, d0.unit()), new Vertex(p1, d1.unit())]));

        if (closed === true) {
          this.polygons.push(new Polygon([vxEnd, new Vertex(p1, vxEnd.normal.clone()), new Vertex(p0, vxEnd.normal.clone())]));
        }
      }
    } else if (radiusEnd === 0) {
      for (let i = 0; i < resolution; i++) {
        let t0 = (i / resolution) * 2 * Math.PI;
        let t1 = (((i + 1) % resolution) / resolution) * 2 * Math.PI;

        let d0 = axisX.times(Math.cos(t0)).plus(axisY.times(Math.sin(t0)));
        let p0 = start.plus(d0.times(radiusStart));
        let d1 = axisX.times(Math.cos(t1)).plus(axisY.times(Math.sin(t1)));
        let p1 = start.plus(d1.times(radiusStart));
        this.polygons.push(new Polygon([vxEnd, new Vertex(p1, d1.unit()), new Vertex(p0, d0.unit())]));

        if (closed === true) {
          this.polygons.push(new Polygon([vxStart, new Vertex(p0, vxStart.normal.clone()), new Vertex(p1, vxStart.normal.clone())]));
        }
      }
    } else {
      for (let i = 0; i < resolution; i++) {
        let t0 = (i / resolution) * 2 * Math.PI;
        let t1 = ((i + 1) / resolution) * 2 * Math.PI;

        let d0 = axisX.times(Math.cos(t0)).plus(axisY.times(Math.sin(t0)));
        let d1 = axisX.times(Math.cos(t1)).plus(axisY.times(Math.sin(t1)));

        let sp0 = start.plus(d0.times(radiusStart));
        let sp1 = start.plus(d1.times(radiusStart));
        let ep0 = end.plus(d0.times(radiusEnd));
        let ep1 = end.plus(d1.times(radiusEnd));

        let a = new Vertex(sp0, d0.unit());
        let b = new Vertex(sp1, d1.unit());
        let c = new Vertex(ep1, d1.unit());
        let d = new Vertex(ep0, d0.unit());

        this.polygons.push(new Polygon([c, b, a]));
        this.polygons.push(new Polygon([d, c, a]));

        if (closed === true) {
          this.polygons.push(new Polygon([vxStart, new Vertex(sp0, vxStart.normal.clone()), new Vertex(sp1, vxStart.normal.clone())]));
          this.polygons.push(new Polygon([vxEnd, new Vertex(ep1, vxEnd.normal.clone()), new Vertex(ep0, vxEnd.normal.clone())]));
        }
      }
    }
  }
}
