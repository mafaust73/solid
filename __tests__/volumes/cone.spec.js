import { Vector } from "../../src/common/vector";
import { Cone } from "../../src/volumes/cone";
import { Wavefront } from "../../src/io/wavefront";
import fs from "fs";

describe("cone", () => {
  test("radiusStart is zero & closed", () => {
    let cone = new Cone(new Vector(0, 0, 0), new Vector(0, 1, 0), 0, 1, true, 3);

    let compare = new Wavefront().write(cone);
    expect(compare).toMatchSnapshot();
  });

  test("radiusStart is zero & open", () => {
    let cone = new Cone(new Vector(0, 0, 0), new Vector(0, 1, 0), 0, 1, false, 3);
    let compare = new Wavefront().write(cone);
    expect(compare).toMatchSnapshot();
  });

  test("radiusEnd is zero & closed", () => {
    let cone = new Cone(new Vector(0, 0, 0), new Vector(0, 1, 0), 1, 0, true, 3);

    let compare = new Wavefront().write(cone);
    expect(compare).toMatchSnapshot();
  });

  test("radiusEnd is zero & open", () => {
    let cone = new Cone(new Vector(0, 0, 0), new Vector(0, 1, 0), 1, 0, false, 3);
    let compare = new Wavefront().write(cone);
    expect(compare).toMatchSnapshot();
  });

  test("closed", () => {
    let cone = new Cone(new Vector(0, 0, 0), new Vector(0, 1, 0), 1, 2, true, 3);

    let compare = new Wavefront().write(cone);
    expect(compare).toMatchSnapshot();
  });

  test("open", () => {
    let cone = new Cone(new Vector(0, 0, 0), new Vector(0, 1, 0), 1, 2, false, 3);

    let compare = new Wavefront().write(cone);
    expect(compare).toMatchSnapshot();
  });
});
