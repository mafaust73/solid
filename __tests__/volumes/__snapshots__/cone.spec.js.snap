// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cone closed 1`] = `
"v -1.7320508075688774 1 -0.9999999999999996
v -0.8660254037844387 0 -0.4999999999999998
v 0 0 1
v 0 1 2
v -1.7320508075688774 1 -0.9999999999999996
v 0 0 1
v 0 0 0
v 0 0 1
v -0.8660254037844387 0 -0.4999999999999998
v 0 1 0
v -1.7320508075688774 1 -0.9999999999999996
v 0 1 2
v 1.7320508075688767 1 -1.0000000000000009
v 0.8660254037844384 0 -0.5000000000000004
v -0.8660254037844387 0 -0.4999999999999998
v -1.7320508075688774 1 -0.9999999999999996
v 1.7320508075688767 1 -1.0000000000000009
v -0.8660254037844387 0 -0.4999999999999998
v 0 0 0
v -0.8660254037844387 0 -0.4999999999999998
v 0.8660254037844384 0 -0.5000000000000004
v 0 1 0
v 1.7320508075688767 1 -1.0000000000000009
v -1.7320508075688774 1 -0.9999999999999996
v 4.898587196589413e-16 1 2
v 2.4492935982947064e-16 0 1
v 0.8660254037844384 0 -0.5000000000000004
v 1.7320508075688767 1 -1.0000000000000009
v 4.898587196589413e-16 1 2
v 0.8660254037844384 0 -0.5000000000000004
v 0 0 0
v 0.8660254037844384 0 -0.5000000000000004
v 2.4492935982947064e-16 0 1
v 0 1 0
v 4.898587196589413e-16 1 2
v 1.7320508075688767 1 -1.0000000000000009
vn -0.8660254037844388 0 -0.49999999999999983
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 0 1
vn 0 0 1
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 0 1
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0.8660254037844384 0 -0.5000000000000004
vn 0.8660254037844384 0 -0.5000000000000004
vn -0.8660254037844388 0 -0.49999999999999983
vn -0.8660254037844388 0 -0.49999999999999983
vn 0.8660254037844384 0 -0.5000000000000004
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 2.4492935982947064e-16 0 1
vn 2.4492935982947064e-16 0 1
vn 0.8660254037844384 0 -0.5000000000000004
vn 0.8660254037844384 0 -0.5000000000000004
vn 2.4492935982947064e-16 0 1
vn 0.8660254037844384 0 -0.5000000000000004
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
f 1//1 2//2 3//3
f 4//4 5//5 6//6
f 7//7 8//8 9//9
f 10//10 11//11 12//12
f 13//13 14//14 15//15
f 16//16 17//17 18//18
f 19//19 20//20 21//21
f 22//22 23//23 24//24
f 25//25 26//26 27//27
f 28//28 29//29 30//30
f 31//31 32//32 33//33
f 34//34 35//35 36//36
"
`;

exports[`cone open 1`] = `
"v -1.7320508075688774 1 -0.9999999999999996
v -0.8660254037844387 0 -0.4999999999999998
v 0 0 1
v 0 1 2
v -1.7320508075688774 1 -0.9999999999999996
v 0 0 1
v 1.7320508075688767 1 -1.0000000000000009
v 0.8660254037844384 0 -0.5000000000000004
v -0.8660254037844387 0 -0.4999999999999998
v -1.7320508075688774 1 -0.9999999999999996
v 1.7320508075688767 1 -1.0000000000000009
v -0.8660254037844387 0 -0.4999999999999998
v 4.898587196589413e-16 1 2
v 2.4492935982947064e-16 0 1
v 0.8660254037844384 0 -0.5000000000000004
v 1.7320508075688767 1 -1.0000000000000009
v 4.898587196589413e-16 1 2
v 0.8660254037844384 0 -0.5000000000000004
vn -0.8660254037844388 0 -0.49999999999999983
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 0 1
vn 0 0 1
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 0 1
vn 0.8660254037844384 0 -0.5000000000000004
vn 0.8660254037844384 0 -0.5000000000000004
vn -0.8660254037844388 0 -0.49999999999999983
vn -0.8660254037844388 0 -0.49999999999999983
vn 0.8660254037844384 0 -0.5000000000000004
vn -0.8660254037844388 0 -0.49999999999999983
vn 2.4492935982947064e-16 0 1
vn 2.4492935982947064e-16 0 1
vn 0.8660254037844384 0 -0.5000000000000004
vn 0.8660254037844384 0 -0.5000000000000004
vn 2.4492935982947064e-16 0 1
vn 0.8660254037844384 0 -0.5000000000000004
f 1//1 2//2 3//3
f 4//4 5//5 6//6
f 7//7 8//8 9//9
f 10//10 11//11 12//12
f 13//13 14//14 15//15
f 16//16 17//17 18//18
"
`;

exports[`cone radiusEnd is zero & closed 1`] = `
"v 0 1 0
v -0.8660254037844387 0 -0.4999999999999998
v 0 0 1
v 0 0 0
v 0 0 1
v -0.8660254037844387 0 -0.4999999999999998
v 0 1 0
v 0.8660254037844384 0 -0.5000000000000004
v -0.8660254037844387 0 -0.4999999999999998
v 0 0 0
v -0.8660254037844387 0 -0.4999999999999998
v 0.8660254037844384 0 -0.5000000000000004
v 0 1 0
v 0 0 1
v 0.8660254037844384 0 -0.5000000000000004
v 0 0 0
v 0.8660254037844384 0 -0.5000000000000004
v 0 0 1
vn 0 1 0
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 0 1
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 1 0
vn 0.8660254037844384 0 -0.5000000000000004
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 1 0
vn 0 0 1
vn 0.8660254037844384 0 -0.5000000000000004
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
f 1//1 2//2 3//3
f 4//4 5//5 6//6
f 7//7 8//8 9//9
f 10//10 11//11 12//12
f 13//13 14//14 15//15
f 16//16 17//17 18//18
"
`;

exports[`cone radiusEnd is zero & open 1`] = `
"v 0 1 0
v -0.8660254037844387 0 -0.4999999999999998
v 0 0 1
v 0 1 0
v 0.8660254037844384 0 -0.5000000000000004
v -0.8660254037844387 0 -0.4999999999999998
v 0 1 0
v 0 0 1
v 0.8660254037844384 0 -0.5000000000000004
vn 0 1 0
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 0 1
vn 0 1 0
vn 0.8660254037844384 0 -0.5000000000000004
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 1 0
vn 0 0 1
vn 0.8660254037844384 0 -0.5000000000000004
f 1//1 2//2 3//3
f 4//4 5//5 6//6
f 7//7 8//8 9//9
"
`;

exports[`cone radiusStart is zero & closed 1`] = `
"v 0 0 0
v 0 1 1
v -0.8660254037844387 1 -0.4999999999999998
v 0 1 0
v -0.8660254037844387 1 -0.4999999999999998
v 0 1 1
v 0 0 0
v -0.8660254037844387 1 -0.4999999999999998
v 0.8660254037844384 1 -0.5000000000000004
v 0 1 0
v 0.8660254037844384 1 -0.5000000000000004
v -0.8660254037844387 1 -0.4999999999999998
v 0 0 0
v 0.8660254037844384 1 -0.5000000000000004
v 0 1 1
v 0 1 0
v 0 1 1
v 0.8660254037844384 1 -0.5000000000000004
vn 0 -1 0
vn 0 0 1
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 -1 0
vn -0.8660254037844388 0 -0.49999999999999983
vn 0.8660254037844384 0 -0.5000000000000004
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 -1 0
vn 0.8660254037844384 0 -0.5000000000000004
vn 0 0 1
vn 0 1 0
vn 0 1 0
vn 0 1 0
f 1//1 2//2 3//3
f 4//4 5//5 6//6
f 7//7 8//8 9//9
f 10//10 11//11 12//12
f 13//13 14//14 15//15
f 16//16 17//17 18//18
"
`;

exports[`cone radiusStart is zero & open 1`] = `
"v 0 0 0
v 0 1 1
v -0.8660254037844387 1 -0.4999999999999998
v 0 0 0
v -0.8660254037844387 1 -0.4999999999999998
v 0.8660254037844384 1 -0.5000000000000004
v 0 0 0
v 0.8660254037844384 1 -0.5000000000000004
v 0 1 1
vn 0 -1 0
vn 0 0 1
vn -0.8660254037844388 0 -0.49999999999999983
vn 0 -1 0
vn -0.8660254037844388 0 -0.49999999999999983
vn 0.8660254037844384 0 -0.5000000000000004
vn 0 -1 0
vn 0.8660254037844384 0 -0.5000000000000004
vn 0 0 1
f 1//1 2//2 3//3
f 4//4 5//5 6//6
f 7//7 8//8 9//9
"
`;
